### PR TITLE
Optimizations for FiberRuntime runloop

### PR DESCRIPTION
**Pull request by kyri-petrou**
This PR contains a few optimizations and micro-optimizations for the FiberRuntime runloop that I've noticed while previously working on #8745 but wanted to optimize them separately. Let's dig into the different things that have been optimized.

#### 1. Checking for messages while running

This PR introduces a wrapper over `ConcurrentLinkedQueue` which has a weakly consistent `unsafeIsEmpty` method for checking if the inbox contains messages. Since we check for new messages before evaluating every single effect in the runloop, the overhead of calling `poll()` is to check whether the queue is empty is non-negligible.

Since adding messages from outside the FiberRuntime is extremely rare (in most cases it's for interruption), I think we can get away with this weakly consistent check in the runLoop itself. Note that we should never rely on the `unsafeIsEmpty` method at any other point.

#### 2. Stack optimizations

The main optimization here is that we avoid setting entries to `null` whenever we "pop" an entry from the stack when we are at the "shallow" part of the stack (idx < 128). The main reasons for this is that we assume that entries that are in the shallow part of the stack are more likely that they'll be replaced automatically as the pointer moves up and down, so we don't need to manually GC them.

The other (micro)optimization regarding the stack is having it initialized when we first start evaluating the effect and avoid repeatedly checking whether it's null during the runloop. Since the `_stack` will be initialized on any kind of effect other than a `Sync` or `Exit`, the only drawback of this is a very small overhead when we fork things like `ZIO.unit`. However, since realistically all effects that are forked will have at least 1 effect that will need to initalize the stack, I think it's better not to initialize it dynamically.

#### 3. Updating `_lastTrace`

Currently, we update `_lastTrace` whenever the current trace is not `null` or `empty`. However, since ZIO's methods require an implicit trace which is propagated to all methods that are called, it's very common for `_lastTrace` to be updated with the same value multiple times. Since reading the variable is _much_ cheaper than writing it, we first check whether the current trace is different than the old one.

I also added some comments in the PR below with some questions / remarks

## Benchmarking results

I only run the `NarrowFlatMap` and `BroadFlatMap` benchmarks using 1 thread, let me know if you think we need to run other benchmarks as well 

TLDR:
- ~25% increase in throughput for "narrow" flatmaps (i.e., the stack doesn't need to resize)
- ~10% increase in throughput for "broad" flatmaps

series/2.x:
```
| Benchmark                                | size | Mode  | Cnt |     Score   |   Error  | Units |
|------------------------------------------|------|-------|-----|-------------|----------|-------|
| BroadFlatMapBenchmark.zioBroadFlatMap    | 20   | thrpt | 5   | 2838.735    | 19.254   | ops/s |
| NarrowFlatMapBenchmark.zioNarrowFlatMap  | 1000 | thrpt | 5   | 66646.224   | 829.889  | ops/s |
```

PR:
```
| Benchmark                               | size | Mode  | Cnt |     Score   |   Error  | Units |
|-----------------------------------------|------|-------|-----|-------------|----------|-------|
| NarrowFlatMapBenchmark.zioBroadFlatMap  | 20   | thrpt | 5   | 3154.488    | 53.074   | ops/s |
| NarrowFlatMapBenchmark.zioNarrowFlatMap | 1000 | thrpt | 5   | 84474.125   | 866.596  | ops/s |
```

### CODE REVIEW

File: core/shared/src/main/scala/zio/internal/FiberRuntime.scala [hunk 1]
```diff
@@ -308,40 +311,28 @@ final class FiberRuntime[E, A](fiberId: FiberId.Runtime, fiberRefs0: FiberRefs,
 
         case FiberMessage.YieldNow =>
 
-        case message =>
```
Comment (kyri-petrou) [inline core/shared/src/main/scala/zio/internal/FiberRuntime.scala:311]: I'm not sure what this case was meant for since all the extensions of `FiberMessage` are included above 🤔
I removed it but let me know if you think we should add it back for some reason

File: core/shared/src/main/scala/zio/internal/FiberRuntime.scala [hunk 2]
```diff
@@ -1418,4 +1421,36 @@ object FiberRuntime {
     }
 
   private val notBlockingOn: () => FiberId = () => FiberId.None
+
+  private final class Inbox {
+    private[this] val queue    = new java.util.concurrent.ConcurrentLinkedQueue[FiberMessage]()
+    private[this] var _isEmpty = true
```
Comment (kyri-petrou) [inline core/shared/src/main/scala/zio/internal/FiberRuntime.scala]: Should this be made volatile? The only scenario that kind of worries me is the following:

1. Thread 1 (runLoop thread) empties the queue (this implies that a message was added right in the previous iteration as well)
2. Thread 1: calls `queue.isEmpty` and sets `_isEmpty` to `true`
3. Thread 2 (external) adds message to queue immediately after
4. Thread 2: sets `_isEmpty = false` but somehow (2) overrides it with `_isEmpty = true`

The chances of this exact sequence of events happening are astronomically small to begin with, but is it something we need to cater for? 🤔

File: core/shared/src/main/scala/zio/internal/FiberRuntime.scala [hunk 3]
```diff
@@ -249,13 +249,15 @@ final class FiberRuntime[E, A](fiberId: FiberId.Runtime, fiberRefs0: FiberRefs,
...
-  ): ZIO.Erased = {
-    var cur = cur0
+  private def drainQueueWhileRunning(cur0: ZIO.Erased): ZIO.Erased = {
+    if (inbox.unsafeIsEmpty) return cur0
 
+    var cur     = cur0
     var message = inbox.poll()
 
+    // Unfortunately we can't avoid the virtual call to `trace` here
+    if (message ne null) updateLastTrace(cur.trace)
```
Comment (kyri-petrou) [inline core/shared/src/main/scala/zio/internal/FiberRuntime.scala:258]: Seems that this was missing (must have missed it when I worked on #8671) and was causing a test to be flaky. This shouldn't add any performance overhead since we're very rarely processing messages from the inbox

**Comment by kyri-petrou** [comment]
@jdegoes Would you be able to review this PR? You might be able to spot some flaws in the optimizations that I might have missed

File: core/shared/src/main/scala/zio/internal/FiberRuntime.scala [hunk 4]
```diff
@@ -836,11 +844,6 @@ final class FiberRuntime[E, A](fiberId: FiberId.Runtime, fiberRefs0: FiberRefs,
     startStackIndex: Int,
     currentDepth: Int
   ): Exit[Any, Any] = {
-    assert(running.get)
```
Comment (kyri-petrou) [inline core/shared/src/main/scala/zio/internal/FiberRuntime.scala:839]: I removed this since we're already checking that we're running prior to calling this method, that way we avoid calling it repeatedly.

Or should it be added back as a safeguard in case we introduce a bug that sets the flag to `false` while the fiber is still running?

File: core/shared/src/main/scala/zio/internal/FiberRuntime.scala [hunk 5]
```diff
@@ -771,9 +764,24 @@ final class FiberRuntime[E, A](fiberId: FiberId.Runtime, fiberRefs0: FiberRefs,
   }
 
   @inline
-  private def popStackFrame(nextStackIndex: Int): Unit = {
-    _stack(nextStackIndex) = null // GC
```
Comment (kyri-petrou) [inline core/shared/src/main/scala/zio/internal/FiberRuntime.scala:775]: The more I think about this, the more I realise that it's a pretty dangerous thing.

What if instead we didn't GC when `nextStackIndex` is below X (perhaps 300 to coincide with trampolining)?

**Comment by jdegoes** [comment]
@kyri-petrou Will do a detailed review tomorrow!

File: core/shared/src/main/scala/zio/internal/FiberRuntime.scala (hunk 5 above)
Comment (kyri-petrou) [reply core/shared/src/main/scala/zio/internal/FiberRuntime.scala:775]: I decided to go with an auto-gc threshold (of 128). Updated the PR description to match the new approach

File: core/shared/src/main/scala/zio/internal/FiberRuntime.scala (hunk 3 above)
Comment (jdegoes) [reply core/shared/src/main/scala/zio/internal/FiberRuntime.scala:258]: 👍

File: core/shared/src/main/scala/zio/internal/FiberRuntime.scala (hunk 1 above)
Comment (jdegoes) [reply core/shared/src/main/scala/zio/internal/FiberRuntime.scala:311]: It was probably legacy from a different time.

File: core/shared/src/main/scala/zio/internal/FiberRuntime.scala [hunk 6]
```diff
@@ -308,40 +311,28 @@ final class FiberRuntime[E, A](fiberId: FiberId.Runtime, fiberRefs0: FiberRefs,
...
-      val stackLength = stack.length
-
-      if (stackLength < size) {
-        val newSize = if ((size & (size - 1)) == 0) size else Integer.highestOneBit(size) << 1
-
-        val newStack = new Array[Continuation](newSize)
+      val newStack = new Array[Continuation](newSize)
 
-        java.lang.System.arraycopy(stack, 0, newStack, 0, _stackSize)
+      java.lang.System.arraycopy(stack, 0, newStack, 0, stackLength)
```
Comment (jdegoes) [inline core/shared/src/main/scala/zio/internal/FiberRuntime.scala:328]: 👍

File: core/shared/src/main/scala/zio/internal/FiberRuntime.scala (hunk 5 above)
Comment (jdegoes) [reply core/shared/src/main/scala/zio/internal/FiberRuntime.scala:775]: Even though I love the performance improvement, I think people will complain if we are holding onto (unnecessary) memory for arbitrary long periods of time.

One possibility is to clear out the entries when the run loop begins, basically by starting at stack index, and nulling until the first null.

This opens the door for making a null array, e.g. `val NullData = Array.fill[AnyRef](...)(null)` and then using the faster `arraycopy` to null out the extra entries.

However, we'd still have the problem of holding onto memory for an indeterminate amount of time.

How much does this single change contribute to the performance improvements?

File: core/shared/src/main/scala/zio/internal/FiberRuntime.scala (hunk 4 above)
Comment (jdegoes) [reply core/shared/src/main/scala/zio/internal/FiberRuntime.scala:839]: I think it only has overhead if assertions are enabled for the JVM (albeit that might be all the time). It's mainly designed for bug detection.

File: core/shared/src/main/scala/zio/internal/FiberRuntime.scala (hunk 5 above)
Comment (kyri-petrou) [reply core/shared/src/main/scala/zio/internal/FiberRuntime.scala:775]: > One possibility is to clear out the entries when the run loop begins, basically by starting at stack index, and nulling until the first null.

If I'm understanding your recommendation correctly, I believe that this is similar to what's currently implemented, but instead clearing out the entries when the runloop starts we do it [when it exits](https://github.com/zio/zio/pull/8800/files#diff-9a664e551dbf1607e888df13c46b57913d52e59dd88d09b160cd238264961cb4R414) (this also means on every yield / async operation or termination).

Effectively we're holding on to objects _unnecessarily_ only while evaluating synchronous effects, which in almost all cases should be a very small period of time, and only for those objects above the `_stackIndex` until the first `null` entry. Once the runloop finishes, the amount of memory we're holding on to will be the same as previously.

> This opens the door for making a null array, e.g. val NullData = Array.fill[AnyRef](...)(null) and then using the faster arraycopy to null out the extra entries.

It's currently done iteratively but I like this recommendation better 👍 

> How much does this single change contribute to the performance improvements?

Between 5-10% of increase in throughput depending on the benchmark.

File: core/shared/src/main/scala/zio/internal/FiberRuntime.scala (hunk 4 above)
Comment (kyri-petrou) [reply core/shared/src/main/scala/zio/internal/FiberRuntime.scala:839]: I'll add it back 👍

Comment (kyri-petrou) [reply core/shared/src/main/scala/zio/internal/FiberRuntime.scala:839]: By the way I just had a look at `assert`; it seems that assertion generation is controlled during compile-time, not at the JVM level. From `assert` scaladoc:

>  A set of `assert` functions are provided for use as a way to document
    and dynamically check invariants in code. Invocations of `assert` can be elided
    at compile time by providing the command line option `-Xdisable-assertions`,
    which raises `-Xelide-below` above `elidable.ASSERTION`, to the `scalac` command.

We should probably use the `-Xdisable-assertions` compiler flag when we generate the published artifacts, but that's probably better done in a separate PR.

File: core/shared/src/main/scala/zio/internal/FiberRuntime.scala (hunk 2 above)
Comment (jdegoes) [reply core/shared/src/main/scala/zio/internal/FiberRuntime.scala]: This can happen and therefore will happen (Murphy's Law of Concurrency), and making it volatile won't help, you'd have to use an atomic integer to track emptiness if you really wanted to fix it (which would probably defeat the optimization).

I am torn on whether to deal with this now, or bite the bullet and do another ticket (that I have yet to write) on creating a highly optimized concurrent mailbox just for fiber runloop.

Comment (kyri-petrou) [reply core/shared/src/main/scala/zio/internal/FiberRuntime.scala]: @jdegoes I see your point! I'll revert this change in the PR since it's now tracked by https://github.com/zio/zio/issues/8807 and give it a go at tackling it separately

File: core/shared/src/main/scala/zio/internal/FiberRuntime.scala (hunk 4 above)
Comment (jdegoes) [reply core/shared/src/main/scala/zio/internal/FiberRuntime.scala:839]: Great idea!

**Review by jdegoes — APPROVED** [review]
Excellent work!
